x-pack/plugins/security/server/lib/authorization/privilege_serializer.ts

```diff
@@ -6,73 +6,73 @@
 
 const featurePrefix = 'feature_';
 const spacePrefix = 'space_';
-const reservedPrivilegeNames = ['all', 'read'];
+const minimumPrivilegeNames = ['all', 'read'];
+const spaceMinimumPrivilegeNames = minimumPrivilegeNames.map(
+  privilegeName => `${spacePrefix}${privilegeName}`
+);
+const deserializeFeaturePrivilegeRegexp = new RegExp(
+  `^${featurePrefix}([a-zA-Z0-9_-]+)\\.([a-zA-Z0-9_-]+)$`
+);
+
+interface FeaturePrivilege {
+  featureId: string;
+  privilege: string;
+}
 
 export class PrivilegeSerializer {
-  public static serializeGlobalReservedPrivilege(privilegeName: string) {
-    if (!reservedPrivilegeNames.includes(privilegeName)) {
-      throw new Error('Unrecognized global reserved privilege');
+  public static isGlobalMinimumPrivilege(privilegeName: string) {
+    return minimumPrivilegeNames.includes(privilegeName);
+  }
+
+  public static isSpaceMinimumPrivilege(privilegeName: string) {
+    return spaceMinimumPrivilegeNames.includes(privilegeName);
+  }
+
+  public static serializeGlobalMinimumPrivilege(privilegeName: string) {
+    if (!minimumPrivilegeNames.includes(privilegeName)) {
+      throw new Error('Unrecognized global minimum privilege');
     }
 
     return privilegeName;
   }
 
-  public static serializeSpaceReservedPrivilege(privilegeName: string) {
-    if (!reservedPrivilegeNames.includes(privilegeName)) {
-      throw new Error('Unrecognized space reserved privilege');
+  public static serializeSpaceMinimumPrivilege(privilegeName: string) {
+    if (!minimumPrivilegeNames.includes(privilegeName)) {
+      throw new Error('Unrecognized space minimum privilege');
     }
 
     return `${spacePrefix}${privilegeName}`;
   }
 
   public static serializeFeaturePrivilege(featureName: string, privilegeName: string) {
-    return `${featurePrefix}${featureName}_${privilegeName}`;
-  }
-
-  public static serializePrivilegeAssignedGlobally(privilege: string) {
-    if (reservedPrivilegeNames.includes(privilege)) {
-      return privilege;
-    }
-
-    return `${featurePrefix}${privilege}`;
+    return `${featurePrefix}${featureName}.${privilegeName}`;
   }
 
-  public static serializePrivilegeAssignedAtSpace(privilege: string) {
-    if (reservedPrivilegeNames.includes(privilege)) {
-      return `${spacePrefix}${privilege}`;
+  public static deserializeFeaturePrivilege(privilege: string): FeaturePrivilege {
+    const match = privilege.match(deserializeFeaturePrivilegeRegexp);
+    if (!match) {
+      throw new Error(`Feature privilege '${privilege}' didn't match pattern`);
```

Comment:
Should we include a potential cause or fix as part of this message?

Reply:
It should only happen if the user is creating Kibana privileges directly using the ES APIs, or we messed something up.